sdk/core/switcher/entry.S

```diff
@@ -597,36 +597,85 @@ exception_entry_asm:
 
 	// Load the interrupted thread's stack pointer into ct0
 	clc                ct0, TrustedStack_offset_csp(csp)
-	// Fetch the base of compartment stack before cincoffset for later
-	// comparison. The subsequent cincoffset could cause the base to change, if
-	// the capability becomes unrepresentable. That would clear the tag, and so
-	// we'd fault later.  Faulting in the switcher will then trigger a forced
-	// unwind.
-	cgetbase           tp, ct0
-	// Allocate space for the register save frame on the stack.  If we didn't
-	// have enough space here, we'll fault setting up the call frame, which
-	// will detect a trap in the switcher and force unwind.
-	cincoffset         ct0, ct0, -(16*8)
 	// See if we can find a handler:
 	clhu               tp, TrustedStack_offset_frameoffset(csp)
 	li                 t1, TrustedStack_offset_frames
 	beq                tp, t1, .Lend_of_stack
 	addi               tp, tp, -TrustedStackFrame_size
+
 	// ctp points to the current available trusted stack frame.
 	cincoffset         ctp, csp, tp
-	// ct1 now contains the export table for the callee
+	// a0 indicates whether we're calling a stackless error handler (0: stack,
+	// 1: stackless)
+	li                 a0, 0
+
+	// Allocate space for the register save frame on the stack.
+	cincoffset         ct0, ct0, -(16*8)
+
+	// WARNING: ENCODING SPECIFIC.
+	// The following depends on the fact that before-the-start values are not
+	// representable in the CHERIoT encoding and so will clear the tag.  If
+	// this property changes then this will need to be replaced by a check that
+	// against the base of the stack.  Note that this check can't be a simple
+	// cgetbase on ct0, because moving the address below the base sufficiently
+	// far that it's out of *representable* bounds will move the reported base
+	// value (base is a displacement from the address).
+	cgettag            t1, ct0
+	// If there isn't enough space on the stack, see if there's a stackless
+	// handler.
+	beqz               t1, .Ltry_stackless_handler
+
 	clc                ct1, TrustedStackFrame_offset_calleeExportTable(ctp)
-	// Reset the export table pointer to point to the *start* of the export
+	// Set the export table pointer to point to the *start* of the export
 	// table.  It will currently point to the entry point that was raised.
 	// TODO: We might want to pass this to the error handler, it might be
```

Comment:
Minor: I think you could lift this common code before the branch. If you need an extra register maybe you could use `a0` which only contains 1 or 0?

Reply:
a0 needs to be live out at the branch that may enter the second path.